lib/charms/mongodb/v0/config_server_interface.py

```diff
@@ -116,35 +133,72 @@ def __init__(
         )
         # TODO Future PRs handle scale down
 
-    def _on_relation_changed(self, event):
+    def _on_relation_changed(self, event) -> None:
         """Starts/restarts monogs with config server information."""
         relation_data = event.relation.data[event.app]
-        if not relation_data.get(KEYFILE_KEY):
+        if not relation_data.get(KEYFILE_KEY) or not relation_data.get(CONFIG_SERVER_DB_KEY):
             event.defer()
             self.charm.unit.status = WaitingStatus("Waiting for secrets from config-server")
             return
 
-        self.update_keyfile(key_file_contents=relation_data.get(KEYFILE_KEY))
+        updated_keyfile = self.update_keyfile(key_file_contents=relation_data.get(KEYFILE_KEY))
+        updated_config = self.update_config_server_db(
+            config_server_db=relation_data.get(CONFIG_SERVER_DB_KEY)
+        )
+
+        # avoid restarting mongos when possible
+        if not updated_keyfile and not updated_config and self.is_mongos_running():
+            return
+
+        # mongos is not available until it is using new secrets
+        logger.info("Restarting mongos with new secrets")
+        self.charm.unit.status = MaintenanceStatus("starting mongos")
+        self.charm.restart_mongos_service()
+
+        # restart on high loaded databases can be very slow (e.g. up to 10-20 minutes).
+        if not self.is_mongos_running():
+            logger.info("mongos has not started, deferring")
+            self.charm.unit.status = WaitingStatus("Waiting for mongos to start")
+            event.defer()
+            return
 
-        # TODO: Follow up PR. Start mongos with the config-server URI
         # TODO: Follow up PR. Add a user for mongos once it has been started
+        self.charm.unit.status = ActiveStatus()
+
+    def is_mongos_running(self) -> bool:
+        """Returns true if mongos service is running."""
+        with MongosConnection(None, f"mongodb://{MONGOS_SOCKET_URI_FMT}") as mongo:
+            return mongo.is_ready
+
+    def update_config_server_db(self, config_server_db) -> bool:
+        """Updates config server str when necessary."""
+        if self.charm.config_server_db == config_server_db:
+            return False
+
+        mongos_config = self.charm.mongos_config
+        mongos_start_args = get_mongos_args(
+            mongos_config, snap_install=True, config_server_db=config_server_db
+        )
+        add_args_to_env("MONGOS_ARGS", mongos_start_args)
+        self.charm.unit_peer_data["config_server_db"] = json.dumps(config_server_db)
```

Comment:
Can we act similarly to `update_keyfile`?
e.g. generate args and compare them to the content of the env file
avoid storing things in peer data to get better reliability.

Reply:
Sure thing boss!